crates/ty_ide/src/docstring.rs

```diff
@@ -153,6 +149,128 @@ fn documentation_trim(docs: &str) -> String {
     output
 }
 
+/// Given a cleaned up docstring, render it to markdown.
+///
+/// This is by no means perfect but it handles a lot of common formats well.
+///
+/// This code currently assumes docstrings are reStructuredText with significant
+/// line-breaks and indentation. Some constructs like *italic*, **bold**, and `inline code`
+/// are shared between the two formats. Other things like _italic_ are only markdown
+/// and should be escaped.
+fn render_markdown(docstring: &str) -> String {
+    // TODO: there is a convention that `singletick` is for items that can
+    // be looked up in-scope while ``multitick`` is for opaque inline code.
+    // While rendering this we should make note of all the `singletick` locations
+    // and (possibly in a higher up piece of logic) try to resolve the names for
+    // cross-linking. (Similar to `TypeDetails` in the type formatting code.)
+    let mut output = String::new();
+    let mut first_line = true;
+    let mut block_indent = 0;
+    let mut in_doctest = false;
+    let mut starting_literal = false;
+    let mut in_literal = false;
+    let mut in_any_code = false;
+    for untrimmed_line in docstring.lines() {
+        // We can assume leading whitespace has been normalized
+        let mut line = untrimmed_line.trim_start_matches(' ');
+        let line_indent = untrimmed_line.len() - line.len();
+
+        // First thing's first, add a newline to start the new line
+        if !first_line {
+            // If we're not in a codeblock, add trailing space to the line to authentically wrap it
```

Comment:
I don't think I understand this. Why is it necessary to add trailing whitespace to any line? This seems very undesired?

Reply:
It's part of the approach I'm taking where I want to preserve the wrapping/indentation of the original docstring.

It prevents e.g. `param1` and `param2` from being merged together as a single line in:

```
'param1' -- The first parameter description
'param2' -- The second parameter description
            This is a continuation of param2 description.
'param3' -- A parameter without type annotation
```